worker/migrationmaster/worker_test.go

```diff
@@ -684,6 +714,153 @@ func (s *Suite) TestExternalControlABORT(c *gc.C) {
+		[]jujutesting.StubCall{
+			{"StreamModelLog", nil},
+			apiOpenControllerCall,
+		},
+	))
+}
+
+func (s *Suite) TestLogTransferErrorOpeningLogDest(c *gc.C) {
+	s.facade.queueStatus(s.makeStatus(coremigration.LOGTRANSFER))
+	s.connection.streamErr = errors.New("tule lake shuffle")
+
+	s.checkWorkerReturns(c, s.connection.streamErr)
+	s.stub.CheckCalls(c, joinCalls(
+		watchStatusLockdownCalls,
+		[]jujutesting.StubCall{
+			{"StreamModelLog", nil},
+			apiOpenControllerCall,
+			openDestLogStreamCall,
+		},
+	))
+}
+
+func (s *Suite) TestLogTransferErrorWriting(c *gc.C) {
+	s.facade.queueStatus(s.makeStatus(coremigration.LOGTRANSFER))
+	s.facade.logMessages = func(d chan<- common.LogMessage) {
+		safeSend(c, d, common.LogMessage{Message: "the go team"})
+	}
+	s.connection.logStream.writeErr = errors.New("bottle rocket")
+	s.checkWorkerReturns(c, s.connection.logStream.writeErr)
+	s.stub.CheckCalls(c, joinCalls(
+		watchStatusLockdownCalls,
+		[]jujutesting.StubCall{
+			{"StreamModelLog", nil},
+			apiOpenControllerCall,
+			openDestLogStreamCall,
+		},
+	))
+	c.Assert(s.connection.logStream.closeCount, gc.Equals, 1)
+}
+
+func (s *Suite) TestLogTransferSendsRecords(c *gc.C) {
+	t1, err := time.Parse("2006-01-02 15:04", "2016-11-28 16:11")
+	c.Assert(err, jc.ErrorIsNil)
+	s.facade.queueStatus(s.makeStatus(coremigration.LOGTRANSFER))
+	messages := []common.LogMessage{
+		{Message: "the go team"},
+		{Message: "joan as police woman"},
+		{
+			Entity:    "the mules",
+			Timestamp: t1,
+			Severity:  "warning",
+			Module:    "this one",
+			Location:  "nearby",
+			Message:   "ham shank",
+		},
+	}
+	s.facade.logMessages = func(d chan<- common.LogMessage) {
+		for _, message := range messages {
+			safeSend(c, d, message)
+		}
+	}
+
+	s.checkWorkerReturns(c, migrationmaster.ErrMigrated)
+	s.stub.CheckCalls(c, joinCalls(
+		watchStatusLockdownCalls,
+		[]jujutesting.StubCall{
+			{"StreamModelLog", nil},
+			apiOpenControllerCall,
+			openDestLogStreamCall,
+			{"facade.SetPhase", []interface{}{coremigration.REAP}},
+			{"facade.Reap", nil},
+			{"facade.SetPhase", []interface{}{coremigration.DONE}},
+		},
+	))
+	c.Assert(s.connection.logStream.written, gc.DeepEquals, []params.LogRecord{
+		{Message: "the go team"},
+		{Message: "joan as police woman"},
+		{
+			Time:     t1,
+			Module:   "this one",
+			Location: "nearby",
+			Level:    "warning",
+			Message:  "ham shank",
+			Entity:   "the mules",
+		},
+	})
+	c.Assert(s.connection.logStream.closeCount, gc.Equals, 1)
+}
+
+func (s *Suite) TestLogTransferReportsProgress(c *gc.C) {
+	s.facade.queueStatus(s.makeStatus(coremigration.LOGTRANSFER))
+	messages := []common.LogMessage{
+		{Message: "captain beefheart"},
+		{Message: "super furry animals"},
+		{Message: "ezra furman"},
+	}
+	s.facade.logMessages = func(d chan<- common.LogMessage) {
+		for _, message := range messages {
+			safeSend(c, d, message)
+			s.clock.WaitAdvance(15*time.Second, coretesting.ShortWait, 1)
```

Comment:
The individual test is reported as taking 0.000s, there's no measurable increase to the suite runtime.